### DISCUSSION TITLE
[node] callbacks and `this` type

### DISCUSSION THREAD

**Discussion by stronny** [DefinitelyTyped/DefinitelyTyped#70764] [_Misc]
So... here's a trivial example:

```js
import {Server} from 'node:http'
new (class {
  foo = 'bar'
  server = new Server(this.serve)
  serve(request, response) {
    console.log(this.foo) // TS thinks it's `(property) (Anonymous class).foo: string`, an easy mistake to make
    response.end(JSON.stringify(this, undefined, '  ')) // guess what "this" is
  }
  start() {this.server.listen(9876, '127.0.0.1')}
})().start()
```

I would assume it isn't impossible to augment callback types with `(this: Type, ...)` signatures so that static checks can actually help prevent uh... happy accidents.

**Comment by Renegade334** [comment]
The problem here is passing references to unbound class methods. Adding a `this` annotation to the Server constructor's listener parameter would not stop you from passing a class method with a different notion of what `this` points to, nor would it alter the TypeScript compiler's treatment of `this` within a class method definition.

See https://nodejs.org/docs/latest/api/events.html#passing-arguments-and-this-to-listeners) for why `this` is the Server instance in your example.

If you want to use class methods that reference `this` as event callbacks, then you'll need to `bind()` them to the class instance when you pass them to the API. Alternatively, you can pass an arrow function created within a scope where `this` points to the intended target, as this will keep the `this` of its original declaration scope when called as an event callback:

```ts
import {Server} from 'node:http'
new (class {
  server: Server
  constructor() {
    const serve = (request, response) => { /* `this` will point to the anonymous class instance here */ }
    this.server = new Server(serve)
  }
  start() { ... }
})().start()
```

**Reply by stronny** [reply]
Yes, I know how to work around it, the simplest is just to do a quick lambda around the method (`(...args) => this.serve(...args)`), this isn't about me being clueless how to proceed.

I admit, I'm a wee bit disappointed with TypeScript once again, but if I acquiesce to its demands and explicitly type my methods with (this: this, ...), it is in fact able to catch the error, like this:

```js
class Server {
	/**@type {string}*/name
	/**@type {(this: Server, num: number) => boolean}*/callback
	constructor(/**@type string*/name, /**@type (this: Server, num: number) => boolean*/callback) {
		this.name = name
		this.callback = callback
	}
	call() {
		console.log(this.callback(123) ? 'yes' : 'no')
	}
}
class Wrapper {
	server = new Server('', this.is_odd) // Type 'Server' is not assignable to type 'this'.
	constructor() {this.server.call()}
	/**@type {(this: this, num: number) => boolean}*/is_odd(num) {
		console.log(`checking for ${this.server.name}`)
		return Boolean(num % 2)
	}
}
new Wrapper()
```

So, typing the callbacks, while not necessarily immediately helpful, still isn't completely meaningless.

**Reply by Renegade334** [reply]
The only reason that's errored there is because you've explicitly added a `this` annotation to the signature of your `is_odd()` class method. Removing it would stop the compiler from reporting an assignability error there. This would not have raised an error in your original example case.

**Reply by stronny** [reply]
Yes, this is what I've meant when I said "if I acquiesce to its demands and explicitly type my methods".

I am able to do it in my own code, but it won't do much if the callbacks don't carry `this` type also.

Let me reiterate.

* For the check to work, both the callback type in my code AND the callback type in the `@types` package need to explicitly type `this`.
* If both are typed, an error is detected. An ideal outcome.
* If yours is typed, but mine isn't, no error is detected, and also nothing bad happens. This is a state that's ready for gradual improvement, as I am able to type my callbacks.
* If mine is typed, but yours isn't, no error is detected. This is a state I can't improve, because I don't control the types in the package.
* If none are typed, this is the current status quo. No error is detected.

I hope this makes sense? Unless you add `this` types to callbacks, there is nothing I can do to make TS check the signatures. If you do add `this` types to callbacks, I can make TS detect errors, if I choose to do so.

**Reply by Renegade334** [reply]
Changing EventEmitter listeners to declare a specific `this` could potentially be experimented with once the outstanding refactors to event emitter types are squared away, but would be a very large change across the library, and I suspect wouldn't be quirk-free.

My gut feeling is that the proposed usage case boils down to "the compiler could warn that X is a problem, if and only if the user has already identified that X is a problem". The intended safeguard against passing unbound class methods requires one to be aware of the problem already, and to modify the method declaration in a non-obvious way in order to have any effect; if that's the case, then instead of that modification being to add a type label, it may as well be a change that solves the problem intrinsically (for example, declaring the callback as a class arrow function property so that `this` stays bound to the class instance, or turning the method into a getter that returns a bound function).

**Reply by stronny** [reply]
I admit, I was under the impression that TS would do this for me, and I'm quite surprised that it doesn't. In light of this newfound knowledge I'm investigating a possibility to create a custom lint rule that will enforce explicit `this` typing for any and all of my class methods universally.

The whole point of linting (I personally don't compile) is to warn the developer of something that they may have missed. Sure, I am aware of this particular quirk while it's fresh in my memory, but I can guarantee I will fall into this trap many times in the future and I want my editor to mark it red for me.

I also accept that what I'm doing isn't widespread, and I'm wondering how others go about it. Relying on memory alone is not a sustainable plan for javascript in my book.

**Reply by Renegade334** [reply]
FWIW, typescript-eslint has [unbound-method](https://typescript-eslint.io/rules/unbound-method/), which should pick up the example cases.

**Reply by stronny** [reply]
Yes, I've already enabled it, and even found room for improvement in a related rule.

https://github.com/typescript-eslint/typescript-eslint/issues/10085
